client/async.go

```diff
@@ -0,0 +1,266 @@
+	a.config.SetDefault(flushIntervalConf, 500*time.Millisecond)
+	a.flushInterval = a.config.GetDuration(flushIntervalConf)
+
+	flushSizeConf := fmt.Sprintf("%sclient.flushSize", configPrefix)
+	a.config.SetDefault(flushSizeConf, 50)
+	a.flushSize = a.config.GetInt(flushSizeConf)
+
+	channelBufferConf := fmt.Sprintf("%sclient.channelBuffer", configPrefix)
+	a.config.SetDefault(channelBufferConf, 200)
+	channelBuffer := a.config.GetInt(channelBufferConf)
+	a.eventsChannel = make(chan *pb.Event, channelBuffer)
+
+	maxRetriesConf := fmt.Sprintf("%sclient.maxRetries", configPrefix)
+	a.config.SetDefault(maxRetriesConf, 3)
+	a.maxRetries = a.config.GetInt(maxRetriesConf)
+
+	retryIntervalConf := fmt.Sprintf("%sclient.retryInterval", configPrefix)
+	a.config.SetDefault(retryIntervalConf, 1*time.Second)
+	a.retryInterval = a.config.GetDuration(retryIntervalConf)
+
+	timeoutConf := fmt.Sprintf("%sclient.grpc.timeout", configPrefix)
+	a.config.SetDefault(timeoutConf, 500*time.Millisecond)
+	a.timeout = a.config.GetDuration(timeoutConf)
+
+	a.logger = a.logger.WithFields(logrus.Fields{
+		"flushInterval": a.flushInterval,
+		"flushSize":     a.flushSize,
+		"channelBuffer": channelBuffer,
+		"timeout":       a.timeout,
+	})
+
+	if err := a.configureGRPCForwarderClient(serverAddress, client); err != nil {
+		return nil, err
+	}
+
+	numRoutinesConf := fmt.Sprintf("%sclient.numRoutines", configPrefix)
+	a.config.SetDefault(numRoutinesConf, 5)
+	numSendRoutines := a.config.GetInt(numRoutinesConf)
+
+	a.logger = a.logger.WithFields(logrus.Fields{
+		"numRoutines": numSendRoutines,
+	})
+
+	for i := 0; i < numSendRoutines; i++ {
+		go a.sendRoutine()
+	}
+
+	return a, nil
+}
+
+func (a *gRPCClientAsync) configureGRPCForwarderClient(
+	serverAddress string,
+	client pb.GRPCForwarderClient,
+	opts ...grpc.DialOption,
+) error {
+	if client != nil {
+		a.client = client
+		return nil
+	}
+	a.logger.WithFields(logrus.Fields{
+		"operation": "configureGRPCForwarderClient",
+	}).Info("connecting to grpc server")
+	dialOpts := append(
+		[]grpc.DialOption{
+			grpc.WithInsecure(),
+			grpc.WithUnaryInterceptor(a.metricsReporterInterceptor),
+		},
+		opts...,
+	)
+	conn, err := grpc.Dial(serverAddress, dialOpts...)
+	if err != nil {
+		return err
+	}
+	a.conn = conn
+	a.client = pb.NewGRPCForwarderClient(conn)
+	return nil
+}
+
+// metricsReporterInterceptor will report metrics from client
+func (a *gRPCClientAsync) metricsReporterInterceptor(
+	ctx context.Context,
+	method string,
+	req interface{},
+	reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	l := a.logger.WithFields(logrus.Fields{
+		"method": method,
+	})
+
+	events := req.(*pb.SendEventsRequest).Events
+
+	defer func(startTime time.Time) {
+		elapsedTime := float64(time.Since(startTime).Nanoseconds() / 1000000)
+		for _, e := range events {
+			metrics.ClientRequestsResponseTime.WithLabelValues(
```

Comment:
@ghostec are we separating sync from async metrics? If not, I think we should send a tag with this info.

Reply:
we can do that by using "route", async calls SendEvents and sync SendEvent

Reply:
nice!